ssh-add.1
.\"  -*- nroff -*-
.\"
.\" ssh-add.1
.\"
.\" Author: Tatu Ylonen <ylo@ssh.fi>
.\"
.\" Copyright (c) 1995 Tatu Ylonen <ylo@ssh.fi>, Espoo, Finland
.\" Copyright (c) 1995-1999 SSH Communications Security Oy, Espoo, Finland
.\"                         All rights reserved
.\"
.\" Created: Sat Apr 22 23:55:14 1995 ylo
.\"
.\" $Id: ssh-add.1,v 1.6 1999/11/17 17:04:55 tri Exp $
.\" $Log: ssh-add.1,v $
.\" Revision 1.6  1999/11/17 17:04:55  tri
.\" 	Fixed copyright notices.
.\"
.\" Revision 1.5  1998/07/08 00:40:34  kivinen
.\" 	Changed to do similar commercial #ifdef processing than other
.\" 	files.
.\"
.\" Revision 1.4  1997/04/27  21:48:43  kivinen
.\" 	Added F-SECURE stuff.
.\"
.\" Revision 1.3  1997/04/23 00:03:50  kivinen
.\" 	Documented -p flag.
.\"
.\" Revision 1.2  1997/03/25 05:41:32  kivinen
.\" 	Changed ylo's email to @ssh.fi.
.\"
.\" Revision 1.1.1.1  1996/02/18 21:38:13  ylo
.\" 	Imported ssh-1.2.13.
.\"
.\" Revision 1.5  1995/08/31  09:22:20  ylo
.\" 	Proofreading fixes.
.\"
.\" Revision 1.4  1995/08/29  22:30:37  ylo
.\" 	Improved manual pages from Andrew Macpherson.
.\"
.\" Revision 1.3  1995/08/21  23:27:06  ylo
.\" 	Minor additions.
.\"
.\" Revision 1.2  1995/07/13  01:36:56  ylo
.\" 	Removed "Last modified" header.
.\" 	Added cvs log.
.\"
.\" $Endlog$
.\"
.\"
.\"
.\"
.\" #ifndef F_SECURE_COMMERCIAL
.TH SSH-ADD 1 "November 8, 1995" "SSH" "SSH"
.\" #endif F_SECURE_COMMERCIAL

.SH NAME
ssh-add \- adds identities for the authentication agent

.SH SYNOPSIS
.B ssh-add
[\c
.B \-p\c
]
[\c
.B \-l\c
]
[\c
.B \-d\c
]
[\c
.B \-D\c
]
[\c
.B file\c
\&.\|.\|.\|]

.SH DESCRIPTION 
.LP
.B Ssh-add
adds identities to the authentication agent,
.B ssh-agent.
When run without arguments, it adds the file
.IR $HOME/\s+2.\s0ssh/identity ".
Alternative file names can be given on the
command line.  If any file requires a passphrase,
.B ssh-add
asks for the passphrase from the user.  If the -p option is given then
the passphrase is read from stdin, otherwise if the user is using X11,
the passphrase is requested using a small X11 program; otherwise it is
read from the user's tty.  (Note: it may be necessary to redirect stdin
from /dev/null to get the passphrase requested using X11.)
.LP
The authentication agent must be running and must be an ancestor of
the current process for
.B ssh-add
to work.

.SH OPTIONS
.TP 0.5i
.B \-p
Read passphrase from stdin (or pipe).
.TP
.B \-l
Lists all identities currently represented by the agent.
.TP
.B \-d
Instead of adding the identity, removes the identity from the agent.
.TP
.B \-D
Deletes all identities from the agent.

.SH RETURN STATUS

.B Ssh-add
returns one of the following exit statuses.  These may be useful in scripts.
.IP 0
The requested operation was performed successfully.
.IP 1
No connection could be made to the authentication agent.  Presumably
there is no authentication agent active in the execution environment
of
.B ssh-add.
.IP 2
The user did not supply a required passphrase.
.IP 3
An identify file could not be found, was not readable, or was in
bad format.
.IP 4
The agent does not have the requested identity.
.IP 5
An unspecified error has occurred; this is a catch-all for errors not
listed above.
.RT

.SH FILES
.TP 
.I \&$HOME/\s+2.\s0ssh/identity
Contains the RSA authentication identity of the user.  This file
should not be readable by anyone but the user.  It is possible to
specify a passphrase when generating the key; that passphrase will be
used to encrypt the private part of this file.  This is the
default file added by
.B ssh-add
when no other files have been specified.
.IP
If
.B ssh-add
needs a passphrase, it will read the passphrase from the current
terminal if it was run from a terminal.  If
.B ssh-add
does not have a terminal associated with it but
.SM DISPLAY\s0
is set, it
will open an X11 window to read the passphrase.  This is particularly
useful when calling
.B ssh-add
from a .Xsession or related script.  (Note that on some machines it
may be necessary to redirect the input from /dev/null to make this work.)

.SH AUTHOR
.LP
Tatu Ylonen <ylo@ssh.fi>

.SH SEE ALSO
.BR ssh-agent (1),
.BR ssh-keygen (1),
.BR ssh (1),
.BR sshd (8)


ssh-add.c
/*

ssh-add.c

Author: Tatu Ylonen <ylo@cs.hut.fi>

Copyright (c) 1995 Tatu Ylonen <ylo@cs.hut.fi>, Espoo, Finland
                   All rights reserved

Created: Thu Apr  6 00:52:24 1995 ylo

Adds an identity to the authentication server, or removes an identity.

*/

/*
 * $Id: ssh-add.c,v 1.3 1995/08/29 22:24:21 ylo Exp $
 * $Log: ssh-add.c,v $
 * Revision 1.3  1995/08/29  22:24:21  ylo
 * 	Added delete_all.
 *
 * Revision 1.2  1995/07/13  01:38:15  ylo
 * 	Removed "Last modified" header.
 * 	Added cvs log.
 *
 * $Endlog$
 */

#include "includes.h"
#include "randoms.h"
#include "rsa.h"
#include "ssh.h"
#include "xmalloc.h"
#include "authfd.h"

void delete_file(const char *filename)
{
  RSAPublicKey key;
  char *comment;
  AuthenticationConnection *ac;

  if (!load_public_key(filename, &key, &comment))
    {
      printf("Bad key file %s: %s\n", filename, strerror(errno));
      return;
    }

  /* Send the request to the authentication agent. */
  ac = ssh_get_authentication_connection();
  if (!ac)
    {
      fprintf(stderr,
	      "Could not open a connection to your authentication agent.\n");
      rsa_clear_public_key(&key);
      xfree(comment);
      return;
    }
  if (ssh_remove_identity(ac, &key))
    fprintf(stderr, "Identity removed: %s (%s)\n", filename, comment);
  else
    fprintf(stderr, "Could not remove identity: %s\n", filename);
  rsa_clear_public_key(&key);
  xfree(comment);
  ssh_close_authentication_connection(ac);
}

void delete_all()
{
  AuthenticationConnection *ac;
  
  /* Get a connection to the agent. */
  ac = ssh_get_authentication_connection();
  if (!ac)
    {
      fprintf(stderr,
	      "Could not open a connection to your authentication agent.\n");
      return;
    }

  /* Send a request to remove all identities. */
  if (ssh_remove_all_identities(ac))
    fprintf(stderr, "All identities removed.\n");
  else
    fprintf(stderr, "Failed to remove all identitities.\n");
  
  /* Close the connection to the agent. */
  ssh_close_authentication_connection(ac);
}

void add_file(const char *filename)
{
  RSAPrivateKey key;
  RSAPublicKey public_key;
  AuthenticationConnection *ac;
  char *saved_comment, *comment;
  
  if (!load_public_key(filename, &public_key, &saved_comment))
    {
      printf("Bad key file %s: %s\n", filename, strerror(errno));
      return;
    }
  
  if (!load_private_key(filename, "", &key, &comment))
    {
      char buf[1024];
      char *pass;
      FILE *f;
      
      printf("Need passphrase for %s (%s).\n", filename, saved_comment);
      if (getenv("DISPLAY") && !isatty(fileno(stdin)))
	{
	  printf("Executing ssh-askpass to query the password...\n");
	  fflush(stdout);
	  fflush(stderr);
	  f = popen("ssh-askpass", "r");
	  if (!fgets(buf, sizeof(buf), f))
	    {
	      pclose(f);
	      xfree(saved_comment);
	      return;
	    }
	  pclose(f);
	  if (strchr(buf, '\n'))
	    *strchr(buf, '\n') = 0;
	  pass = xstrdup(buf);
	}
      else
	{
	  sprintf(buf, "Enter passphrase: ");
	  pass = read_passphrase(buf, 1);
	}
	  
      if (!load_private_key(filename, pass, &key, &comment))
	{
	  memset(pass, 0, strlen(pass));
	  fprintf(stderr, 
		  "Bad passphrase, file not readable, or invalid keyfile.\n");
	  xfree(saved_comment);
	  exit(1);
	}
      memset(pass, 0, strlen(pass));
    }

  xfree(saved_comment);

  /* Send the key to the authentication agent. */
  ac = ssh_get_authentication_connection();
  if (!ac)
    {
      fprintf(stderr,
	      "Could not open a connection to your authentication agent.\n");
      rsa_clear_private_key(&key);
      xfree(comment);
      return;
    }
  if (ssh_add_identity(ac, &key, comment))
    fprintf(stderr, "Identity added: %s (%s)\n", filename, comment);
  else
    fprintf(stderr, "Could not add identity: %s\n", filename);
  rsa_clear_private_key(&key);
  xfree(comment);
  ssh_close_authentication_connection(ac);
}

void list_identities()
{
  AuthenticationConnection *ac;
  MP_INT e, n;
  int bits, status;
  char *comment;
  int had_identities;

  ac = ssh_get_authentication_connection();
  if (!ac)
    {
      fprintf(stderr, "Could not connect to authentication server.\n");
      return;
    }
  mpz_init(&e);
  mpz_init(&n);
  had_identities = 0;
  for (status = ssh_get_first_identity(ac, &bits, &e, &n, &comment);
       status;
       status = ssh_get_next_identity(ac, &bits, &e, &n, &comment))
    {
      had_identities = 1;
      printf("%d ", bits);
      mpz_out_str(stdout, 10, &e);
      printf(" ");
      mpz_out_str(stdout, 10, &n);
      printf(" %s\n", comment);
      xfree(comment);
    }
  mpz_clear(&e);
  mpz_clear(&n);
  if (!had_identities)
    printf("The agent has no identities.\n");
  ssh_close_authentication_connection(ac);
}

int main(int ac, char **av)
{
  struct passwd *pw;
  char buf[1024];
  int no_files = 1;
  int i;
  int deleting = 0;

  for (i = 1; i < ac; i++)
    {
      if (strcmp(av[i], "-l") == 0)
	{
	  list_identities();
	  no_files = 0; /* Don't default-add/delete if -l. */
	  continue;
	}
      if (strcmp(av[i], "-d") == 0)
	{
	  deleting = 1;
	  continue;
	}
      if (strcmp(av[i], "-D") == 0)
	{
	  delete_all();
	  no_files = 0;
	  continue;
	}
      no_files = 0;
      if (deleting)
	delete_file(av[i]);
      else
	add_file(av[i]);
    }
  if (no_files)
    {
      pw = getpwuid(getuid());
      if (!pw)
	{
	  fprintf(stderr, "No user found with uid %d\n", (int)getuid());
	  exit(1);
	}
      sprintf(buf, "%s/%s", pw->pw_dir, SSH_CLIENT_IDENTITY);
      if (deleting)
	delete_file(buf);
      else
	add_file(buf);
    }
  exit(0);
}

void fatal(const char *fmt, ...)
{
  va_list args;
  va_start(args, fmt);
  vfprintf(stderr, fmt, args);
  fprintf(stderr, "\n");
  va_end(args);
  exit(1);
}
